platform/include/platform/CircularBuffer.h

```diff
@@ -84,77 +86,195 @@ class CircularBuffer {
+     * @param src Data to be pushed to the buffer.
      */
-    bool pop(T &data)
+    void push(mbed::Span<const T> src)
+    {
+        push(src.data(), src.size());
+    }
+
+    /** Pop from the buffer.
+     *
+     * @param data Container to store the data to be popped from the buffer.
+     * @return True if data popped.
+     */
+    bool pop(T& data)
     {
         bool data_popped = false;
+
         core_util_critical_section_enter();
-        if (!empty()) {
-            data = _pool[_tail++];
-            if (_tail == BufferSize) {
-                _tail = 0;
+
+        if (!non_critical_empty()) {
+            data_popped = true;
+
+            data = _buffer[_tail];
+            _tail = incrementCounter(_tail);
+            _full = false;
+        }
+
+        core_util_critical_section_exit();
+
+        return data_popped;
+    }
+
+    /**
+     * Pop multiple elements from the buffer.
+     *
+     * @param dest The array which will receive the elements.
+     * @param len The number of elements to pop.
+     *
+     * @return The number of elements popped.
+     */
+    CounterType pop(T* dest, CounterType len)
+    {
+        if (len == 0) {
+            return 0;
+        }
+
+        CounterType data_popped = 0;
+
+        core_util_critical_section_enter();
+
+        if(!non_critical_empty()) {
+            /* make sure we only try to read as much as we have items present */
+            if (len > non_critical_size()) {
+                len = non_critical_size();
             }
+            data_popped = len;
+
+            /* items may be split by overlap, take only the number we have to the right of tail */
+            if ((_tail + data_popped) > BufferSize) {
+                data_popped = BufferSize - _tail;
+            }
+
+            std::copy(_buffer + _tail, _buffer + _tail + data_popped, dest);
+
+            /* if we looped over the end we may need to pop again */
+            CounterType left_to_pop = len - data_popped;
+
+            if (left_to_pop) {
+                std::copy(_buffer, _buffer + left_to_pop, dest + data_popped);
+
+                data_popped += left_to_pop;
+            }
+
+            _tail += data_popped;
+            _tail %= BufferSize;
             _full = false;
-            data_popped = true;
         }
+
         core_util_critical_section_exit();
+
         return data_popped;
     }
 
-    /** Check if the buffer is empty
+    /**
+     * Pop multiple elements from the buffer.
+     *
+     * @param dest The span that contains the buffer that will be used to store the elements.
      *
-     * @return True if the buffer is empty, false if not
+     * @return The span with the size set to number of elements popped using the buffer passed in as the parameter.
+     */
+    mbed::Span<T> pop(mbed::Span<T> dest)
+    {
+        CounterType popped = pop(dest.data(), dest.size());
+        return mbed::make_Span(dest.data(), popped);
```

Comment:
Just a random note that as of current C++ version you can do `return {dest.data(), popped};`.

Don't know what our style view on that is. I'm conflicted.

Reply:
I think it's great to push good C++ coding standard.

https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md uses `return {dest.data(), popped};` so I think it's good to stick with that.

Reply:
Do you have a more specific reference? That's quite a big page, and can't see anything specifically discussing this case.

Reply:
They don't discuss, they use the construct in code examples for `std::pair` [here](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#example-284) -- they don't use `std::make_pair`

Reply:
Then it's not really a specific style judgment on how explicit or not to be, which is what I'd like to see.

And there's something of a difference - a pair is a pair, so it's "just" a tuple-type container. You're effectively filling in an aggregate from a list, which is the oldest-school use of `{}`.

Here, you're invoking a constructor. I always feel a bit icky when using `{}` on its own to invoke a constructor. Maybe I'm too old-fashioned. At least it's the most "direct" constructor. Doing `return { start_ptr, end_ptr };` to use the iterator constructor would feel weirder.

Anyway, I'm broadly in favour of the more concise style; I don't like excess clutter. I think the case for it is even stronger in function parameter context, where the span could just be one of several arguments. Whereas on `return` you do usually have space to spare for that little type reminder.

(C++17 will change the situation again, because you no longer have to use `make_Span`, you can have the deduction guides for the constructor letting you do just `Span(...)`, so the clutter of explicitness is reduced further.)

Reply:
Thanks for the detailed explanations :) (your comments are always great learning material for me)

One side question (also asked [here](https://forums.mbed.com/t/std-span-or-mbed-span/10175), what's the difference between `mbed::Span` and `std::span`? Which one should I use in my application code? and why the `S` for `Span`?

Reply:
`mbed::Span` is based on an old revision of the `std::span` proposal, the only difference - API wise - with the one in the (next!) standard is the type used to return the size of the span: it is signed in mbed, not signed on std. I believe we should make the switch to `unsigned` at some point. Another difference is the debug capability of the iterator, it is very limited in mbed whereas libc++ returns _fat_ iterators in debug.

Reply:
Just seen Vincent's answer - this partly repeats it...

`std::span` is C++20, so not supported by all toolchains we support, so can't use it directly. We're also in C++14 mode, and it may not be accessible in that mode.

`mbed::Span` was written and added to mbed while C++20 was in draft, so has some minor differences compared to the final revision of `std::span` - in particular its use of `ptrdiff_t` instead of `size_t`. The capital `S` was to fit mbed style guidelines.

Now that C++20 has been finalised, and following other work done in this style pending full `std` availability, I would suggest implementing and migrating to a new `mstd::span` that directly matches `std::span` alongside the other `mstd` stuff in the cxxsupport directory.

The `mstd` types like `mstd::mutex` and `mstd::atomic` are close enough matches to `std` that they're drop-in replacements, possibly just needing to change 1 `using` directive and 1 include.

Reply:
Thank you both (again) for the detailed answers :)

I was not aware of the [`mstd`](https://github.com/ARMmbed/mbed-os/tree/master/platform/cxxsupport#c-support)! This is great stuff!